### PR TITLE
DEV: moves logic from job to a service

### PR DESCRIPTION
**Pull request by jjaffeux**
`Jobs::AutoJoinChannelBatch` was holding a lot of logic which should be in a service. Moreover, this refactoring is the opportunity to address a bug which could cause a duplicate key error.

From now when trying to insert a new membership it won't fail if a membership is already present.

Example error:

```
Job exception: ERROR:  duplicate key value violates unique constraint "user_chat_channel_unique_memberships"
DETAIL:  Key (user_id, chat_channel_id)=(1, 2) already exists.

Backtrace
rack-mini-profiler-3.1.0/lib/patches/db/pg.rb:110:in `exec' 
rack-mini-profiler-3.1.0/lib/patches/db/pg.rb:110:in `async_exec' 
(eval):29:in `async_exec'
mini_sql-1.4.0/lib/mini_sql/postgres/connection.rb:209:in `run' 
mini_sql-1.4.0/lib/mini_sql/active_record_postgres/connection.rb:38:in `block in run' 
mini_sql-1.4.0/lib/mini_sql/active_record_postgres/connection.rb:34:in `block in with_lock' 
activesupport-7.0.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt' 
activesupport-7.0.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize' 
activesupport-7.0.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt' 
activesupport-7.0.5.1/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize' 
mini_sql-1.4.0/lib/mini_sql/active_record_postgres/connection.rb:34:in `with_lock' 
mini_sql-1.4.0/lib/mini_sql/active_record_postgres/connection.rb:38:in `run' 
mini_sql-1.4.0/lib/mini_sql/postgres/connection.rb:64:in `query_single' 
/var/www/discourse/plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb:38:in `execute' 
```

### CODE REVIEW

File: plugins/chat/app/jobs/regular/auto_join_channel_batch.rb [hunk 1]
```diff
@@ -1,79 +0,0 @@
-# NOTE: When changing auto-join logic, make sure to update the `settings.auto_join_users_info` translation as well.
```
Comment (jjaffeux) [inline plugins/chat/app/jobs/regular/auto_join_channel_batch.rb:1]: This whole file should have been removed during the zeitwerk move

File: plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb [hunk 2]
```diff
@@ -1,81 +1,10 @@
...
-            (last_seen_at IS NULL OR last_seen_at > :last_seen_at) AND
-            uo.chat_enabled AND
-            uccm.id IS NULL AND
-            (NOT EXISTS(SELECT 1 FROM category_groups WHERE category_id = :channel_category) 
-              OR EXISTS (SELECT 1 FROM category_groups WHERE category_id = :channel_category AND group_id = :everyone AND permission_type <= :permission_type)
-              OR cg.category_id = :channel_category)
-  
-          RETURNING user_chat_channel_memberships.user_id
-        SQL
+        ::Chat::AutoJoinChannelBatch.call(**args)
```
Comment (jjaffeux) [inline plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb:7]: I consider logic in services as harmful as logic in controllers. It might even be more harmful due to the asynchronous nature of jobs making side effects even harder to track.

File: plugins/chat/app/services/chat/auto_join_channel_batch.rb [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            NOT EXISTS(SELECT 1 FROM anonymous_users a WHERE a.user_id = users.id) AND
+            (suspended_till IS NULL OR suspended_till <= :suspended_until) AND
+            (last_seen_at IS NULL OR last_seen_at > :last_seen_at) AND
+            uo.chat_enabled AND
+
+            (NOT EXISTS(SELECT 1 FROM category_groups WHERE category_id = :channel_category)
+              OR EXISTS (SELECT 1 FROM category_groups WHERE category_id = :channel_category AND group_id = :everyone AND permission_type <= :permission_type)
+              OR cg.category_id = :channel_category)
+
+          ON CONFLICT DO NOTHING
```
Comment (jjaffeux) [inline plugins/chat/app/services/chat/auto_join_channel_batch.rb:106]: this is the fix essentially

File: plugins/chat/spec/jobs/regular/chat/auto_join_channel_batch_spec.rb [hunk 4]
```diff
@@ -1,253 +0,0 @@
-# frozen_string_literal: true
```
Comment (jjaffeux) [inline plugins/chat/spec/jobs/regular/chat/auto_join_channel_batch_spec.rb:1]: other specs are checking that their actions are enqueing this job and/or are seeing the effects of it

File: plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb (hunk 2 above)
Comment (Flink) [reply plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb:7]: > consider logic in **services** as harmful as logic in controllers.

I think you meant in **jobs** :grin:

Comment (jjaffeux) [reply plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb:7]: yes indeed :p

Comment (martin-brennan) [inline plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb:7]: Hmm I thought we were saying that services shouldn't be called from jobs? What if there is a failure in the service? The job will succeed as if everything is fine. We should at the least have some capture for the generic `on_failure` that fails the job with whatever reason that the service failed, no?

So something like:

```ruby
result = ::Chat::AutoJoinChannelBatch.call(**args)
if result.failure?
  raise StandardError.new(result.errors)
end
```

I still don't think we have an easy way to just spit out the reasons why a service failed; we have inspect_steps but that just shows the passing/failing steps. Maybe we need something like this for this job use case? @Flink what are your thoughts here?

File: plugins/chat/app/services/chat/auto_join_channel_batch.rb [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+    contract
+    model :channel
+    model :category
+    step :create_memberships
+    step :recalculate_user_count
+    step :publish_new_channel
+
+    class Contract
+      attribute :chat_channel_id, :integer
```
Comment (martin-brennan) [inline plugins/chat/app/services/chat/auto_join_channel_batch.rb:25]: Can we just change to `channel_id` now?

File: plugins/chat/app/services/chat/auto_join_channel_batch.rb [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      validates :chat_channel_id, presence: true
+      attribute :starts_at, :integer
+      attribute :ends_at, :integer
+
+      validate :boundaries
+      def boundaries
+        if %w[starts_at ends_at].all? { |attr| self.public_send(attr).blank? }
```
Comment (martin-brennan) [inline plugins/chat/app/services/chat/auto_join_channel_batch.rb:32]: The error is saying `starts_at OR ends_at missing` so this should be `any?` not `all?`

File: plugins/chat/app/services/chat/auto_join_channel_batch.rb [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+          errors.add :base, "starts_at or ends_at missing"
+        end
+
+        if ends_at && starts_at
+          errors.add(:base, "ends_at can't be higher than starts_at") if ends_at < starts_at
```
Comment (martin-brennan) [inline plugins/chat/app/services/chat/auto_join_channel_batch.rb:37]: This is backwards too, should be `if ends_at > starts_at`

File: plugins/chat/app/services/chat/auto_join_channel_batch.rb [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+    end
+
+    def fetch_category(channel:, **)
+      channel.chatable
+    end
+
+    def create_memberships(channel:, contract:, category:, **)
+      query_args = {
+        chat_channel_id: channel.id,
+        start: contract.starts_at,
```
Comment (martin-brennan) [inline plugins/chat/app/services/chat/auto_join_channel_batch.rb:55]: I know these start/end params are from the old service but probably makes sense to rename them as `start_user_id` and `end_user_id`, otherwise on first glance it seems like they should be start/end dates.

File: plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb (hunk 2 above)
Comment (jjaffeux) [reply plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb:7]: Yes I thought of this while working on it, and I just think we need the same thing than what we have for controllers @Flink 

Besides this "lack" there's no reason to not have services in jobs.

File: plugins/chat/app/services/chat/auto_join_channel_batch.rb (hunk 6 above)
Comment (jjaffeux) [reply plugins/chat/app/services/chat/auto_join_channel_batch.rb:32]: Yes maybe I will check, I don't think I changed the existing logic here though, but something is probably no correct

File: plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb (hunk 2 above)
Comment (Flink) [reply plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb:7]: I’m used to just call a service/class/whatever from a job and not have logic directly in it so no problem in that regard from me :grin: 
And we already have a special job class to use our services in jobs: https://github.com/discourse/discourse/blob/main/plugins/chat/app/jobs/service_job.rb
I remember doing this after discussing a bit in a PR, but it seems we didn’t use the class in the end :sweat_smile:

Comment (jjaffeux) [reply plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb:7]: nice, I also thought we already spoke of this indeed, ok cool, I will use it thanks

File: plugins/chat/app/services/chat/auto_join_channel_batch.rb (hunk 5 above)
Comment (Flink) [reply plugins/chat/app/services/chat/auto_join_channel_batch.rb:25]: We could probably do something like this while migrating:
```ruby
attribute :chat_channel_id, :integer
attribute :channel_id, :integer

validates :channel_id, presence: true

before_validation { self.channel_id ||= chat_channel_id }
```
And then we’ll just have to remove this extra piece of code later. Wdyt?

Comment (jjaffeux) [reply plugins/chat/app/services/chat/auto_join_channel_batch.rb:25]: Yes I just didn't want to add this on top of the rest, but we can do this :p

File: plugins/chat/app/services/chat/auto_join_channel_batch.rb (hunk 7 above)
Comment (jjaffeux) [reply plugins/chat/app/services/chat/auto_join_channel_batch.rb:37]: no I think this is the text which is incorrect

File: plugins/chat/app/services/chat/auto_join_channel_batch.rb (hunk 8 above)
Comment (jjaffeux) [reply plugins/chat/app/services/chat/auto_join_channel_batch.rb:55]: Yes this is also something I did at first :p

File: plugins/chat/spec/services/chat/auto_join_channel_batch_spec.rb [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+  subject(:result) { described_class.call(params) }
+
+  let(:chat_channel_id) { channel_1.id }
+  let(:starts_at) { 0 }
+  let(:ends_at) { 10 }
+  let(:params) { { chat_channel_id: chat_channel_id, starts_at: starts_at, ends_at: ends_at } }
+
+  fab!(:channel_1) { Fabricate(:chat_channel, auto_join_users: true) }
+
+  context "when contract" do
```
[the remark this answers is not in the document]
Comment (Flink) [reply plugins/chat/spec/services/chat/auto_join_channel_batch_spec.rb:15]: I’m just missing specs for the new action, I’m putting what I have already here:

_spec/services/chat/auto_join_channel_batch_spec.rb_
```ruby
# frozen_string_literal: true

require "rails_helper"

describe Chat::AutoJoinChannelBatch do
  describe Chat::AutoJoinChannelBatch::Contract, type: :model do
    subject(:contract) { described_class.new(start_user_id: 10) }

    it { is_expected.to validate_presence_of(:channel_id) }
    it { is_expected.to validate_presence_of(:start_user_id) }
    it { is_expected.to validate_presence_of(:end_user_id) }
    it { is_expected.to validate_comparison_of(:end_user_id).is_greater_than(:start_user_id) }

    describe "Backward compatibility" do
      subject(:contract) { described_class.new(args) }

      before { contract.valid? }

      context "when providing 'chat_channel_id'" do
        let(:args) { { chat_channel_id: 2 } }

        it "sets 'channel_id'" do
          expect(contract.channel_id).to eq(2)
        end
      end

      context "when providing 'starts_at'" do
        let(:args) { { starts_at: 5 } }

        it "sets 'start_user_id'" do
          expect(contract.start_user_id).to eq(5)
        end
      end

      context "when providing 'ends_at'" do
        let(:args) { { ends_at: 8 } }

        it "sets 'end_user_id'" do
          expect(contract.end_user_id).to eq(8)
        end
      end
    end
  end

  describe ".call" do
    subject(:result) { described_class.call(params) }

    fab!(:channel) { Fabricate(:chat_channel, auto_join_users: true) }

    let(:channel_id) { channel.id }
    let(:start_user_id) { 0 }
    let(:end_user_id) { 10 }
    let(:params) do
      { channel_id: channel_id, start_user_id: start_user_id, end_user_id: end_user_id }
    end

    context "when arguments are invalid" do
      let(:channel_id) { nil }

      it { is_expected.to fail_a_contract }
    end

    context "when arguments are valid" do
      context "when channel does not exist" do
        let(:channel_id) { -1 }

        it { is_expected.to fail_to_find_a_model(:channel) }
      end

      context "when channel is not a category channel" do
        fab!(:channel) { Fabricate(:direct_message_channel, auto_join_users: true) }

        it { is_expected.to fail_to_find_a_model(:channel) }
      end

      context "when channel is not in auto_join_users mode" do
        before { channel.update!(auto_join_users: false) }

        it { is_expected.to fail_to_find_a_model(:channel) }
      end

      context "when channel is found" do
        fab!(:users) { Fabricate.times(2, :user) }

        let(:manager) { mock.responds_like_instance_of(Chat::ChannelMembershipManager) }

        before do
          Chat::Action::CreateMembershipsForAutoJoin
            .stubs(:call)
            .with(has_entries(channel: channel, contract: instance_of(described_class::Contract)))
            .returns(user_ids)
          Chat::ChannelMembershipManager.stubs(:new).with(channel).returns(manager)
          manager.stubs(:recalculate_user_count)
        end

        context "when more than one membership is created" do
          let(:user_ids) { users.map(&:id) }

          it "does not recalculate user count" do
            manager.expects(:recalculate_user_count).never
            result
          end

          it "publishes an event" do
            Chat::Publisher.expects(:publish_new_channel).with(channel, users)
            result
          end
        end

        context "when only one membership is created" do
          let(:user_ids) { [users.first.id] }

          it "recalculates user count" do
            manager.expects(:recalculate_user_count)
            result
          end

          it "publishes an event" do
            Chat::Publisher.expects(:publish_new_channel).with(channel, [users.first])
            result
          end
        end
      end
    end
  end
end
```
_app/services/chat/auto_join_channel_batch.rb_
```ruby
# NOTE: When changing auto-join logic, make sure to update the `settings.auto_join_users_info` translation as well.
# frozen_string_literal: true

module Chat
  # Service responsible to create memberships for a channel and a section of user ids
  #
  # @example
  #  Chat::AutoJoinChannelBatch.call(
  #   channel_id: 1,
  #   start_user_id: 27,
  #   end_user_id: 58,
  #  )
  #
  class AutoJoinChannelBatch
    include Service::Base

    contract
    model :channel
    step :create_memberships
    step :recalculate_user_count
    step :publish_new_channel

    class Contract
      # Backward-compatible attributes
      attribute :chat_channel_id, :integer
      attribute :starts_at, :integer
      attribute :ends_at, :integer

      # New attributes
      attribute :channel_id, :integer
      attribute :start_user_id, :integer
      attribute :end_user_id, :integer

      validates :channel_id, :start_user_id, :end_user_id, presence: true
      validates :end_user_id, comparison: { greater_than: :start_user_id }

      # TODO: remove after migration is done
      before_validation do
        self.channel_id ||= chat_channel_id
        self.start_user_id ||= starts_at
        self.end_user_id ||= ends_at
      end
    end

    private

    def fetch_channel(contract:, **)
      ::Chat::CategoryChannel.find_by(id: contract.channel_id, auto_join_users: true)
    end

    def create_memberships(channel:, contract:, **)
      context.added_user_ids =
        Chat::Action::CreateMembershipsForAutoJoin.call(channel: channel, contract: contract)
    end

    def recalculate_user_count(channel:, added_user_ids:, **)
      # Only do this if we are running auto-join for a single user, if we
      # are doing it for many then we should do it after all batches are
      # complete for the channel in Jobs::AutoJoinChannelMemberships
      return unless added_user_ids.one?
      ::Chat::ChannelMembershipManager.new(channel).recalculate_user_count
    end

    def publish_new_channel(channel:, added_user_ids:, **)
      ::Chat::Publisher.publish_new_channel(channel.reload, User.where(id: added_user_ids))
    end
  end
end
```
_app/services/chat/action/create_memberships_for_auto_join.rb_
```ruby
# frozen_string_literal: true

module Chat
  module Action
    class CreateMembershipsForAutoJoin
      def self.call(channel:, contract:)
        query_args = {
          chat_channel_id: channel.id,
          start: contract.start_user_id,
          end: contract.end_user_id,
          suspended_until: Time.zone.now,
          last_seen_at: 3.months.ago,
          channel_category: channel.category.id,
          permission_type: CategoryGroup.permission_types[:create_post],
          everyone: Group::AUTO_GROUPS[:everyone],
          mode: ::Chat::UserChatChannelMembership.join_modes[:automatic],
        }
        ::DB.query_single(<<~SQL, query_args)
          INSERT INTO user_chat_channel_memberships (user_id, chat_channel_id, following, created_at, updated_at, join_mode)
          SELECT DISTINCT(users.id), :chat_channel_id, TRUE, NOW(), NOW(), :mode
          FROM users
          INNER JOIN user_options uo ON uo.user_id = users.id
          LEFT OUTER JOIN user_chat_channel_memberships uccm ON
            uccm.chat_channel_id = :chat_channel_id AND uccm.user_id = users.id

          LEFT OUTER JOIN group_users gu ON gu.user_id = users.id
          LEFT OUTER JOIN category_groups cg ON cg.group_id = gu.group_id AND
          cg.permission_type <= :permission_type

          WHERE (users.id >= :start AND users.id <= :end) AND
            users.staged IS FALSE AND
            users.active AND
            NOT EXISTS(SELECT 1 FROM anonymous_users a WHERE a.user_id = users.id) AND
            (suspended_till IS NULL OR suspended_till <= :suspended_until) AND
            (last_seen_at IS NULL OR last_seen_at > :last_seen_at) AND
            uo.chat_enabled AND

            (NOT EXISTS(SELECT 1 FROM category_groups WHERE category_id = :channel_category)
              OR EXISTS (SELECT 1 FROM category_groups WHERE category_id = :channel_category AND group_id = :everyone AND permission_type <= :permission_type)
              OR cg.category_id = :channel_category)

          ON CONFLICT DO NOTHING

          RETURNING user_chat_channel_memberships.user_id
        SQL
      end
    end
  end
end
```

[the remark this answers is not in the document]
Comment (jjaffeux) [reply plugins/chat/spec/services/chat/auto_join_channel_batch_spec.rb:15]: Yep nice :+1: I considered writing an action for the insert too but decided against it for now,, but agreed it's the best :+1: and now that's it's done 😅

[the remark this answers is not in the document]
Comment (Flink) [reply plugins/chat/spec/services/chat/auto_join_channel_batch_spec.rb:15]: And here is the spec for the new action (mostly it’s just what you wrote previously):
```ruby
# frozen_string_literal: true

RSpec.describe Chat::Action::CreateMembershipsForAutoJoin do
  subject(:action) { described_class.call(channel: channel, contract: contract) }

  fab!(:channel) { Fabricate(:chat_channel, auto_join_users: true) }
  fab!(:user_1) { Fabricate(:user, last_seen_at: 15.minutes.ago) }

  let(:start_user_id) { user_1.id }
  let(:end_user_id) { user_1.id }
  let(:contract) { OpenStruct.new(start_user_id: start_user_id, end_user_id: end_user_id) }

  it "adds correct members" do
    expect(action).to eq([user_1.id])
  end

  it "sets the reason to automatic" do
    action
    expect(channel.membership_for(user_1)).to be_automatic
  end

  context "with others users not in the batch" do
    fab!(:user_2) { Fabricate(:user) }

    it "adds correct members" do
      expect(action).to eq([user_1.id])
    end
  end

  context "with suspended users" do
    before { user_1.update!(suspended_till: 1.year.from_now) }

    it "skips suspended users" do
      expect(action).to eq([])
    end
  end

  context "with users not seen recently" do
    before { user_1.update!(last_seen_at: 4.months.ago) }

    it "skips users last_seen more than 3 months ago" do
      expect(action).to eq([])
    end
  end

  context "with never seen users" do
    before { user_1.update!(last_seen_at: nil) }

    it "skips users with last_seen set to null" do
      expect(action).to eq([user_1.id])
    end
  end

  context "with disabled chat users" do
    before { user_1.user_option.update!(chat_enabled: false) }

    it "skips users without chat_enabled" do
      expect(action).to eq([])
    end
  end

  context "with anonymous users" do
    fab!(:user_1) { Fabricate(:anonymous, last_seen_at: 15.minutes.ago) }

    it "skips anonymous users" do
      expect(action).to eq([])
    end
  end

  context "with inactive users" do
    before { user_1.update!(active: false) }

    it "skips inactive users" do
      expect(action).to eq([])
    end
  end

  context "with staged users" do
    before { user_1.update!(staged: true) }

    it "skips staged users" do
      expect(action).to eq([])
    end
  end

  context "when user is already a member" do
    before { channel.add(user_1) }

    it "is a noop" do
      expect(action).to eq([])
    end
  end

  context "when category is restricted" do
    fab!(:user_1) { Fabricate(:user) }
    fab!(:user_2) { Fabricate(:user) }
    fab!(:group_1) { Fabricate(:group) }
    fab!(:channel) { Fabricate(:private_category_channel, group: group_1, auto_join_users: true) }

    let(:end_user_id) { user_2.id }

    before { group_1.add(user_1) }

    it "only joins users with access to the category through the group" do
      expect(action).to eq([user_1.id])
    end

    context "when the user has access through multiple groups" do
      fab!(:group_2) { Fabricate(:group) }

      before do
        channel.category.category_groups.create!(
          group_id: group_2.id,
          permission_type: CategoryGroup.permission_types[:full],
        )
        group_2.add(user_1)
      end

      it "correctly joins the user" do
        expect(action).to eq([user_1.id])
      end
    end

    context "when the category group is read only" do
      fab!(:channel) { Fabricate(:private_category_channel, auto_join_users: true) }

      before do
        channel.category.category_groups.create!(
          group_id: group_1.id,
          permission_type: CategoryGroup.permission_types[:readonly],
        )
        group_1.add(user_1)
      end

      it "doesn’t join the users of the group" do
        expect(action).to eq([])
      end
    end

    context "when the category group has create post permission" do
      fab!(:channel) { Fabricate(:private_category_channel, auto_join_users: true) }

      before do
        channel.category.category_groups.create!(
          group_id: group_1.id,
          permission_type: CategoryGroup.permission_types[:create_post],
        )
        group_1.add(user_1)
      end

      it "correctly joins the user" do
        expect(action).to eq([user_1.id])
      end
    end

    context "when user has allowed groups and disallowed groups" do
      fab!(:group_2) { Fabricate(:group) }

      before do
        channel.category.category_groups.create!(
          group_id: group_2.id,
          permission_type: CategoryGroup.permission_types[:readonly],
        )
        group_2.add(user_1)
      end

      it "correctly joins the user" do
        expect(action).to eq([user_1.id])
      end
    end
  end
end
```
There’s something weird though:
```ruby
  context "with never seen users" do
    before { user_1.update!(last_seen_at: nil) }

    it "skips users with last_seen set to null" do
      expect(action).to eq([user_1.id])
    end
  end
```
This one is green so it might be a bug? :thinking:

[the remark this answers is not in the document]
Comment (Flink) [reply plugins/chat/spec/services/chat/auto_join_channel_batch_spec.rb:15]: Just remembered that we have to use the `main` branch of `shoulda-matchers` to have access to the `validate_comparison_of` matcher (I guess they will release a new version soon).

File: plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb (hunk 2 above)
Comment (martin-brennan) [reply plugins/chat/app/jobs/regular/chat/auto_join_channel_batch.rb:7]: Wow what a shit memory I have, thanks for that @Flink  :laughing:

**Comment by Drenmi** [comment]
I definitely agree with this. To me a job is just an entry point, same as a controller action, and I am used to just calling a service from inside. The responsibility of the job becomes to deal with failures, retries, etc. 👌 Some benefits is service is easier to test in isolation, and can be used in other call sites.

**Comment by jjaffeux** [comment]
@martin-brennan @Flink I will merge this tomorrow unless one of you stops me :p

File: Gemfile [hunk 10]
```diff
@@ -158,7 +158,7 @@ group :test, :development do
 
   gem "rspec-rails"
 
-  gem "shoulda-matchers", require: false
+  gem "shoulda-matchers", require: false, github: "thoughtbot/shoulda-matchers"
```
Comment (martin-brennan) [inline Gemfile:161]: Why is this necessary? I see it's for `greater_than_or_equal_to` but have they not cut a new gem recently that includes that or something? We should probably have a `TODO` to go back and change this to the non-github version at some point.

**Comment by martin-brennan** [comment]
> I will merge this tomorrow unless one of you stops me

Go ahead...make my day :rocket:

File: Gemfile (hunk 10 above)
Comment (jjaffeux) [reply Gemfile:161]: Yep no gem AFAICT, I make a todo topic

File: plugins/chat/spec/services/actions/create_memberships_for_auto_join_spec.rb [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    it "skips users last_seen more than 3 months ago" do
+      expect(action).to eq([])
+    end
+  end
+
+  context "with never seen users" do
+    before { user_1.update!(last_seen_at: nil) }
+
+    it "skips users with last_seen set to null" do
+      expect(action).to eq([user_1.id])
```
Comment (Flink) [inline plugins/chat/spec/services/actions/create_memberships_for_auto_join_spec.rb:50]: Is it a bug or not in the end? We’re saying it should skip users with `last_seen_at` set to null but that’s not the case it seems :sweat_smile:

File: plugins/chat/spec/jobs/regular/chat/auto_join_channel_batch_spec.rb [hunk 12]
```diff
@@ -3,251 +3,30 @@
...
-
-        other_group = Fabricate(:group)
-        Fabricate(
-          :category_group,
-          category: restricted_category,
-          group: other_group,
-          permission_type: CategoryGroup.permission_types[:create_post],
-        )
-        other_group.add(another_user)
+  it "logs an error on failed contract" do
```
Comment (Flink) [inline plugins/chat/spec/jobs/regular/chat/auto_join_channel_batch_spec.rb:17]: Usually, when referring to a condition in an `it`, it’s a hint to use a `context` :grin: 
I’d rewrite this like that instead:
```ruby
context "when contract fails" do
  before { Jobs.run_immediately! }

  it "logs an error" do
    …
  end
end
```

File: plugins/chat/spec/jobs/regular/chat/auto_join_channel_batch_spec.rb [hunk 13]
```diff
@@ -3,251 +3,30 @@
...
-    end
+    Jobs.enqueue(described_class)
   end
 
-  def assert_users_follows_channel(channel, users)
-    new_memberships = Chat::UserChatChannelMembership.where(user: users, chat_channel: channel)
-    expect(new_memberships.length).to eq(users.length)
-    expect(new_memberships.all?(&:following)).to eq(true)
-  end
+  it "logs an error on failed model" do
```
Comment (Flink) [inline plugins/chat/spec/jobs/regular/chat/auto_join_channel_batch_spec.rb:25]: Same here:
```ruby
context "when model is not found" do
  before { Jobs.run_immediately! }

  it "logs an error" do
    …
  end
end
```

**Review by Flink — COMMENTED** [review]
Just a question about a potential bug, otherwise really tiny things :+1:

File: plugins/chat/spec/jobs/regular/chat/auto_join_channel_batch_spec.rb (hunk 13 above)
Comment (jjaffeux) [reply plugins/chat/spec/jobs/regular/chat/auto_join_channel_batch_spec.rb:25]: Yes I wanted to avoid so many lines for simple cases but agreed :+1:

File: plugins/chat/spec/services/actions/create_memberships_for_auto_join_spec.rb (hunk 11 above)
Comment (jjaffeux) [reply plugins/chat/spec/services/actions/create_memberships_for_auto_join_spec.rb:50]: will check :+1:

Comment (jjaffeux) [reply plugins/chat/spec/services/actions/create_memberships_for_auto_join_spec.rb:50]: Yep I checked, the text of the test is wrong (my fault) and it should correctly include this user

Comment (Flink) [reply plugins/chat/spec/services/actions/create_memberships_for_auto_join_spec.rb:50]: Perfect then! :partying_face:

**Review by Flink — APPROVED** [review]
:rocket:
